docs/enhancements/0001-worker-pods.md

```diff
@@ -0,0 +1,223 @@
+That filesystem disappears as soon as the Pod is terminated.  
+This means that node image caching mechanisms cannot be leveraged, and any Pod creation will entail data transfers from
+the kmod image's registry to the Pod.
+
+##### Setting the firmware lookup path
+
+Worker Pods will be able to write the firmware lookup path to `/sys/module/firmware_class/parameters/path` each time
+they run.
+This behaviour will be controlled by an operator setting will be written as part of the worker configuration.
+
+#### High-level reconciliation loop
+
+```mermaid
+flowchart TD
+    M[Module update]
+    N[Node update]
+    NSM[NodeModuleState\nspec update]
+    P[Worker Pod starts\nrunning on Node]
+    DMO{NodeModuleState\nspec matches status?}
+    E[End]
+    
+    M --> NSM
+    N --> NSM
+    
+    NSM --> DMO
+    
+    DMO -- Yes --> E
+    DMO -- No --> P
+    
+    CR[Controller starts] --> P
+    NR[Node becomes Ready] --> P
+    
+    P -- Pod completed or failed --> RO[Controller reads\nPod output] --> OUS[Controller updates\nNodeModuleState status] --> DMO
+```
+
+#### Device plugins
+
+KMM will still run the device plugin specified in the `Module` resource as a DaemonSet; however instead of relying on
+the `kmm.node.kubernetes.io/$namespace.$module.ready` label in its selector, we will introduce a new label
+that is independent of the kmod's state in the node.
+This should make it easier to terminate the device plugin before we try to unload the kmod.
+
+#### Addressing goals
+
+This design addresses the goals stated at the beginning of this proposal:
+
+1. **Improve the reliability of module loading and unloading**  
+   KMM now runs its own code in the pod, which gives us a lot more control in error handling.
+
+2. **Improve operator upgrade experience**  
+   Worker pods are one-off, short-lived objects.
+   We never try to update them and are not constrained by spec changes and immutable selectors.
+
+3. **Reduce the overall KMM footprint**  
+   We now run only one Pod per Node, for only a short amount of time.
+
+### KMM-Hub
+
+KMM-Hub is not affected by this change.
+
+## API changes
+
+We do not foresee any change to the existing APIs, including `Module`, for this enhancement.  
+The new `ModuleNodeState` CRD will be strictly internal to KMM and should not be relied upon for any purpose.  
+Kernel version labeling for nodes will be dropped as it is not required to schedule ModuleLoader Pods anymore.
+
+## Upgrading from DaemonSet-based KMM
+
+There is no practical way of deleting the existing ModuleLoader DaemonSets without running their Pods' `preStop` hook.
+That hook runs `modprobe -r` and unloads the kernel module(s).  
+This means that transitioning from DaemonSets to the new system will entail an interruption of the workload utilizing
+the kernel module.  
+To minimize the downtime, we may add an additional controller that leverages the DaemonSet `OnDelete` strategy to delete
```

Comment:
Another way is to ask customer to provide and executable that take a reference to the kernel module. We will deploy it as one-time DS,and it will allow us to remove ModuleLoader DS without worrying that we will unload the kernel module

Reply:
In fact I think we could maybe increment the counter ourselves? There's `try_module_get`, although it seems [deprecated](https://stackoverflow.com/questions/1741415/linux-kernel-modules-when-to-use-try-module-get-module-put).

Reply:
i looked into yet, but i could not find the common way that will suit all the kernel modules. I think it is OK to ask customer to provide such an executable, since he is familiar with the code and must know how to do it ( if it is needed at all)

Reply:
Added a few lines about that idea.